br/pkg/pdutil/pd.go

```diff
@@ -834,125 +576,47 @@ func (p *PdController) doRemoveSchedulersWith(
+	return p.pauseSchedulersAndConfigWith(ctx, needRemoveSchedulers, disablePDCfg)
 }
 
 // GetMinResolvedTS get min-resolved-ts from pd
 func (p *PdController) GetMinResolvedTS(ctx context.Context) (uint64, error) {
-	var err error
-	for _, addr := range p.getAllPDAddrs() {
-		v, e := pdRequest(ctx, addr, pdhttp.MinResolvedTSPrefix, p.cli, http.MethodGet, nil)
-		if e != nil {
-			log.Warn("failed to get min resolved ts", zap.String("addr", addr), zap.Error(e))
-			err = e
-			continue
-		}
-		log.Info("min resolved ts", zap.String("resp", string(v)))
-		d := struct {
-			IsRealTime    bool   `json:"is_real_time,omitempty"`
-			MinResolvedTS uint64 `json:"min_resolved_ts"`
-		}{}
-		err = json.Unmarshal(v, &d)
-		if err != nil {
-			return 0, errors.Trace(err)
-		}
-		if !d.IsRealTime {
-			message := "min resolved ts not enabled"
-			log.Error(message, zap.String("addr", addr))
-			return 0, errors.Trace(errors.New(message))
-		}
-		return d.MinResolvedTS, nil
-	}
-	return 0, errors.Trace(err)
+	ts, _, err := p.pdHTTPCli.GetMinResolvedTSByStoresIDs(ctx, nil)
+	return ts, errors.Trace(err)
 }
 
 // RecoverBaseAllocID recover base alloc id
 func (p *PdController) RecoverBaseAllocID(ctx context.Context, id uint64) error {
-	reqData, _ := json.Marshal(&struct {
-		ID string `json:"id"`
-	}{
-		ID: fmt.Sprintf("%d", id),
-	})
-	var err error
-	for _, addr := range p.getAllPDAddrs() {
-		_, e := pdRequest(ctx, addr, pdhttp.BaseAllocID, p.cli, http.MethodPost, reqData)
-		if e != nil {
-			log.Warn("failed to recover base alloc id", zap.String("addr", addr), zap.Error(e))
-			err = e
-			continue
-		}
-		return nil
-	}
-	return errors.Trace(err)
+	return p.pdHTTPCli.ResetBaseAllocID(ctx, id)
 }
 
 // ResetTS reset current ts of pd
 func (p *PdController) ResetTS(ctx context.Context, ts uint64) error {
 	// reset-ts of PD will never set ts < current pd ts
 	// we set force-use-larger=true to allow ts > current pd ts + 24h(on default)
-	reqData, _ := json.Marshal(&struct {
-		Tso            string `json:"tso"`
-		ForceUseLarger bool   `json:"force-use-larger"`
-	}{
-		Tso:            fmt.Sprintf("%d", ts),
-		ForceUseLarger: true,
-	})
-	var err error
-	for _, addr := range p.getAllPDAddrs() {
-		code, _, e := pdRequestWithCode(ctx, addr, pdhttp.ResetTS, p.cli, http.MethodPost, reqData)
-		if e != nil {
-			// for pd version <= 6.2, if the given ts < current ts of pd, pd returns StatusForbidden.
-			// it's not an error for br
-			if code == http.StatusForbidden {
-				log.Info("reset-ts returns with status forbidden, ignore")
-				return nil
-			}
-			log.Warn("failed to reset ts", zap.Uint64("ts", ts), zap.String("addr", addr), zap.Error(e))
-			err = e
-			continue
-		}
+	err := p.pdHTTPCli.ResetTS(ctx, ts, true)
+	if err == nil {
+		return nil
+	}
+	if strings.Contains(err.Error(), http.StatusText(http.StatusForbidden)) {
+		log.Info("reset-ts returns with status forbidden, ignore")
 		return nil
 	}
-	return errors.Trace(err)
+	return err
```

Comment:
seems some path of pd http client doesn't contains stack, so keep `errors.Trace(err)`?